examples/reshare/src/dkg/manager.rs

```diff
@@ -0,0 +1,643 @@
+                    shares,
+                    acks,
+                    outcome: meta.local_outcome,
+                })
+            } else {
+                None
+            }
+        } else {
+            share.as_ref().map(|share| {
+                let (dealer, commitment, shares) =
+                    Dealer::new(context, Some(share.clone()), players.clone());
+                DealerMetadata {
+                    dealer,
+                    commitment,
+                    shares,
+                    acks: BTreeMap::new(),
+                    outcome: None,
+                }
+            })
+        };
+
+        let (s, r) = mux.register(epoch as u32).await.unwrap();
+
+        Self {
+            signer,
+            epoch,
+            previous: share.map_or(RoundResult::Polynomial(public.clone()), |share| {
+                RoundResult::Output(Output { public, share })
+            }),
+            players,
+            sender: s,
+            receiver: r,
+            dealer_meta,
+            player,
+            arbiter,
+            store,
+        }
+    }
+
+    /// Distribute the [Dealer]'s 'shares to all contributors that have not yet acknowledged receipt of their share.
+    pub async fn distribute(&mut self, round: u64) {
+        // Only attempt distribution if the manager is also a dealer.
+        let Some(DealerMetadata {
+            dealer,
+            commitment,
+            shares,
+            acks,
+            ..
+        }) = &mut self.dealer_meta
+        else {
+            return;
+        };
+
+        // Find all contributors that need to be sent a share by filtering out those that have
+        // not yet acknowledged receipt of their share.
+        let needs_broadcast = self
+            .players
+            .iter()
+            .enumerate()
+            .filter(|(i, _)| !acks.contains_key(&(*i as u32)))
+            .collect::<Vec<_>>();
+
+        for (idx, contributor) in needs_broadcast {
+            let share = shares.get(idx).cloned().unwrap();
+
+            if let Some((signer_index, ref mut player)) = self.player {
+                if idx == signer_index as usize {
+                    player
+                        .share(
+                            self.signer.public_key(),
+                            commitment.deref().clone(),
+                            share.clone(),
+                        )
+                        .unwrap();
+                    dealer.ack(self.signer.public_key()).unwrap();
+
+                    let ack = Ack::new::<_, V>(
+                        ACK_NAMESPACE,
+                        self.signer,
+                        signer_index,
+                        round,
+                        &self.signer.public_key(),
+                        commitment,
+                    );
+                    acks.insert(signer_index, ack.clone());
+
+                    // Persist the acknowledgement to storage.
+                    self.store
```

Comment:
If you are using `upsert` like this, I think `Metadata` is a better fit (will be more efficient).

I think the right way to use `store` would be `key = <epoch>-<type> ...`.

Reply:
Upsert is useful 👍